### ISSUE TITLE
Unzip error after version 3.1.0

### ISSUE THREAD

**Issue by yhon-acurio** [nodeca/probe-image-size#16]
# Description
I upgrade from 3.0.0 to 3.1.0 and I get a new error related to unzip operations. Do you know if there is a regression linked to new dependencies?
I noticed, unzip-response is not used anymore in 3.1.0 and my exception is linked to unzip operations.

I attach logs.

# Comparing dependencies
```
+-- probe-image-size@3.0.0 |   | 350 | +-- probe-image-size@3.1.0
-- | -- | -- | --
  | 352 | \| +-- any-promise@1.3.0 |   | 351 | \| +-- any-promise@1.3.0
  | 353 | \| +-- deepmerge@1.3.2 |   | 352 | \| +-- deepmerge@1.5.0
  | 354 | \| +-- got@6.7.1 |   | 353 | \| +-- got@7.1.0
  | 355 | \| \| +-- create-error-class@3.0.2 |   | 354 | \| \| +-- decompress-response@3.3.0
  | 356 | \| \| \| `-- capture-stack-trace@1.0.0 |   | 355 | \| \| \| `-- mimic-response@1.0.0
  | 357 | \| \| +-- duplexer3@0.1.4 |   | 356 | \| \| +-- duplexer3@0.1.4
  | 358 | \| \| +-- get-stream@3.0.0 |   | 357 | \| \| +-- get-stream@3.0.0
  | 359 | \| \| +-- is-redirect@1.0.0 |   | 358 | \| \| +-- is-plain-obj@1.1.0
  | 360 | \| \| +-- is-retry-allowed@1.1.0 |   | 359 | \| \| +-- is-retry-allowed@1.1.0
  | 361 | \| \| +-- is-stream@1.1.0 |   | 360 | \| \| +-- is-stream@1.1.0
  |   |   |   | 361 | \| \| +-- isurl@1.0.0
  |   |   |   | 362 | \| \| \| +-- has-to-string-tag-x@1.4.0
  |   |   |   | 363 | \| \| \| \| `-- has-symbol-support-x@1.4.0
  |   |   |   | 364 | \| \| \| `-- is-object@1.0.1
  | 362 | \| \| +-- lowercase-keys@1.0.0 |   | 365 | \| \| +-- lowercase-keys@1.0.0
  |   |   |   | 366 | \| \| +-- p-cancelable@0.3.0
  |   |   |   | 367 | \| \| +-- p-timeout@1.2.0
  |   |   |   | 368 | \| \| \| `-- p-finally@1.0.0
  | 363 | \| \| +-- timed-out@4.0.1 |   | 369 | \| \| +-- timed-out@4.0.1
  | 364 | \| \| +-- unzip-response@2.0.1 |   |   |  
  | 365 | \| \| `-- url-parse-lax@1.0.0 |   | 370 | \| \| +-- url-parse-lax@1.0.0
  | 366 | \| \|   `-- prepend-http@1.0.4 |   | 371 | \| \| \| `-- prepend-http@1.0.4
  |   |   |   | 372 | \| \| `-- url-to-options@1.0.1
  | 367 | \| +-- inherits@2.0.3 |   | 373 | \| +-- inherits@2.0.3
  | 368 | \| +-- next-tick@1.0.0 |   | 374 | \| +-- next-tick@1.0.0
  | 369 | \| `-- stream-parser@0.3.1
```

# Logs

## In 3.0.0
```
2017-08-14T18:26:17.401978+00:00 app[web.1]:     at DuplexWrapper.<anonymous> (/app/node_modules/probe-image-size/http.js:50:16)
2017-08-14T18:26:17.401979+00:00 app[web.1]:     at emitThree (events.js:121:20)
2017-08-14T18:26:17.401980+00:00 app[web.1]:     at DuplexWrapper.emit (events.js:194:7)
2017-08-14T18:26:17.401981+00:00 app[web.1]:     at EventEmitter.ee.on.res (/app/node_modules/got/index.js:250:10)
2017-08-14T18:26:17.401982+00:00 app[web.1]:     at emitOne (events.js:96:13)
2017-08-14T18:26:17.401982+00:00 app[web.1]:     at EventEmitter.emit (events.js:188:7)
2017-08-14T18:26:17.401983+00:00 app[web.1]:     at Immediate.setImmediate (/app/node_modules/got/index.js:99:8)
2017-08-14T18:26:17.401987+00:00 app[web.1]:     at Immediate.wrapped (/app/node_modules/newrelic/lib/transaction/tracer/index.js:184:28)
2017-08-14T18:26:17.401988+00:00 app[web.1]:     at runCallback (timers.js:649:20)
2017-08-14T18:26:17.401988+00:00 app[web.1]:     at tryOnImmediate (timers.js:622:5)
2017-08-14T18:26:17.401989+00:00 app[web.1]:     at processImmediate [as _immediateCallback] (timers.js:594:5)
```

## In 3.1.0
```
2017-08-14T18:26:17.401978+00:00 app[web.1]:     at DuplexWrapper.<anonymous> (/app/node_modules/probe-image-size/http.js:50:16)
2017-08-14T18:26:17.401979+00:00 app[web.1]:     at emitThree (events.js:121:20)
2017-08-14T18:26:17.401980+00:00 app[web.1]:     at DuplexWrapper.emit (events.js:194:7)
2017-08-14T18:26:17.401981+00:00 app[web.1]:     at EventEmitter.ee.on.res (/app/node_modules/got/index.js:250:10)
2017-08-14T18:26:17.401982+00:00 app[web.1]:     at emitOne (events.js:96:13)
2017-08-14T18:26:17.401982+00:00 app[web.1]:     at EventEmitter.emit (events.js:188:7)
2017-08-14T18:26:17.401983+00:00 app[web.1]:     at Immediate.setImmediate (/app/node_modules/got/index.js:99:8)
2017-08-14T18:26:17.401987+00:00 app[web.1]:     at Immediate.wrapped (/app/node_modules/newrelic/lib/transaction/tracer/index.js:184:28)
2017-08-14T18:26:17.401988+00:00 app[web.1]:     at runCallback (timers.js:649:20)
2017-08-14T18:26:17.401988+00:00 app[web.1]:     at tryOnImmediate (timers.js:622:5)
2017-08-14T18:26:17.401989+00:00 app[web.1]:     at processImmediate [as _immediateCallback] (timers.js:594:5)
2017-08-14T18:26:17.411162+00:00 app[web.1]: Mon Aug 14 2017 18:26:17 GMT+0000 (UTC) - error: app.js uncaughtException App will be exit(1) Error: write after end
2017-08-14T18:26:17.411165+00:00 app[web.1]:     at writeAfterEnd (_stream_writable.js:193:12)
2017-08-14T18:26:17.411166+00:00 app[web.1]:     at Unzip.Writable.write (_stream_writable.js:244:5)
2017-08-14T18:26:17.411167+00:00 app[web.1]:     at IncomingMessage.ondata (_stream_readable.js:555:20)
2017-08-14T18:26:17.411168+00:00 app[web.1]:     at emitOne (events.js:96:13)
2017-08-14T18:26:17.411168+00:00 app[web.1]:     at IncomingMessage.emit (events.js:188:7)
2017-08-14T18:26:17.411169+00:00 app[web.1]:     at IncomingMessage.wrapped (/app/node_modules/newrelic/lib/transaction/tracer/index.js:184:28)
2017-08-14T18:26:17.411170+00:00 app[web.1]:     at IncomingMessage.wrappedResponseEmit [as emit] (/app/node_modules/newrelic/lib/transaction/tracer/instrumentation/outbound.js:99:26)
2017-08-14T18:26:17.411171+00:00 app[web.1]:     at readableAddChunk (_stream_readable.js:176:18)
2017-08-14T18:26:17.411171+00:00 app[web.1]:     at PassThrough.Readable.push (_stream_readable.js:134:10)
2017-08-14T18:26:17.411171+00:00 app[web.1]:     at PassThrough.Transform.push (_stream_transform.js:128:32)
2017-08-14T18:26:17.411172+00:00 app[web.1]:     at afterTransform (_stream_transform.js:77:12)
2017-08-14T18:26:17.411173+00:00 app[web.1]:     at TransformState.afterTransform (_stream_transform.js:54:12)
2017-08-14T18:26:17.411173+00:00 app[web.1]:     at PassThrough._transform (_stream_passthrough.js:21:3)
2017-08-14T18:26:17.411173+00:00 app[web.1]:     at PassThrough.Transform._read (_stream_transform.js:167:10)
2017-08-14T18:26:17.411174+00:00 app[web.1]:     at PassThrough.Transform._write (_stream_transform.js:155:12)
2017-08-14T18:26:17.411174+00:00 app[web.1]:     at doWrite (_stream_writable.js:334:12)
2017-08-14T18:26:17.411175+00:00 app[web.1]:     at writeOrBuffer (_stream_writable.js:320:5)
2017-08-14T18:26:17.411175+00:00 app[web.1]:     at PassThrough.Writable.write (_stream_writable.js:247:11)
2017-08-14T18:26:17.411175+00:00 app[web.1]:     at Unzip.ondata (_stream_readable.js:555:20)
2017-08-14T18:26:17.411176+00:00 app[web.1]:     at emitOne (events.js:96:13)
2017-08-14T18:26:17.411176+00:00 app[web.1]:     at Unzip.emit (events.js:188:7)
2017-08-14T18:26:17.411176+00:00 app[web.1]:     at readableAddChunk (_stream_readable.js:176:18)
2017-08-14T18:26:17.411177+00:00 app[web.1]:     at Unzip.Readable.push (_stream_readable.js:134:10)
2017-08-14T18:26:17.411177+00:00 app[web.1]:     at Unzip.Transform.push (_stream_transform.js:128:32)
2017-08-14T18:26:17.411177+00:00 app[web.1]:     at Zlib.callback (zlib.js:598:14)
```

**Comment by puzrin** [comment]
https://github.com/nodeca/probe-image-size/blob/master/CHANGELOG.md#310--2017-06-08 the only difference is `got` upgrade. Code of package itself was not changed anyhow.

Please provide a minimal sample how to reproduce.

**Comment by puzrin** [comment]
Closed: not a bug, no feedback.

**Comment by puzrin** [comment]
```sh
$ node ./support/dump.js https://raw.githubusercontent.com/fontello/fontello/master/fontello-image.svg
https://raw.githubusercontent.com/fontello/fontello/master/fontello-image.svg: { width: 300,
  height: 194.673,
  type: 'svg',
  mime: 'image/svg+xml',
  wUnits: 'px',
  hUnits: 'px',
  length: 2806,
  url: 'https://raw.githubusercontent.com/fontello/fontello/master/fontello-image.svg' }
```
Works as expected (svg src has gzip header)

**Comment by wong2** [comment]
I've got the same error:

```
Error: write after end
  File "_stream_writable.js", line 220, col 12, in writeAfterEnd
  File "_stream_writable.js", line 271, col 5, in Unzip.Writable.write
  File "_stream_readable.js", line 628, col 20, in IncomingMessage.ondata
  File "events.js", line 115, col 13, in emitOne
  File "events.js", line 210, col 7, in IncomingMessage.emit
  File "/usr/src/app/node_modules/newrelic/lib/transaction/tracer/index.js", line 183, col 28, in IncomingMessage.wrapped
    var result = handler.apply(this, arguments)
  File "/usr/src/app/node_modules/newrelic/lib/transaction/tracer/instrumentation/outbound.js", line 97, col 26, in IncomingMessage.wrappedResponseEmit [as emit]
    return boundEmit.apply(this, arguments)
  File "_stream_readable.js", line 252, col 12, in addChunk
  File "_stream_readable.js", line 239, col 11, in readableAddChunk
  File "_stream_readable.js", line 197, col 10, in PassThrough.Readable.push
  File "_stream_transform.js", line 147, col 32, in PassThrough.Transform.push
  File "_stream_transform.js", line 88, col 10, in PassThrough.afterTransform
  File "_stream_passthrough.js", line 42, col 3, in PassThrough._transform
  File "_stream_transform.js", line 186, col 10, in PassThrough.Transform._read
  File "_stream_transform.js", line 174, col 12, in PassThrough.Transform._write
  File "_stream_writable.js", line 371, col 12, in doWrite
  File "_stream_writable.js", line 357, col 5, in writeOrBuffer
  File "_stream_writable.js", line 274, col 11, in PassThrough.Writable.write
  File "_stream_readable.js", line 628, col 20, in Unzip.ondata
  File "events.js", line 115, col 13, in emitOne
  File "events.js", line 210, col 7, in Unzip.emit
  File "_stream_readable.js", line 252, col 12, in addChunk
  File "_stream_readable.js", line 239, col 11, in readableAddChunk
  File "_stream_readable.js", line 197, col 10, in Unzip.Readable.push
  File "_stream_transform.js", line 147, col 32, in Unzip.Transform.push
  File "zlib.js", line 447, col 14, in Zlib.callback
```

but I haven''t found a url to reproduce this

**Comment by puzrin** [comment]
I would suggest ask in `got` tracker.

**Comment by wong2** [comment]
@puzrin you're right, I've opened an issue there: https://github.com/sindresorhus/got/issues/388

**Comment by puzrin** [comment]
Thanks! I hope got will be fixed soon, it's mainteined well.

**Comment by wong2** [comment]
@puzrin I got a new related error with 3.0.0

```
const probe = require('probe-image-size')

probe('https://mp.weixin.qq.com/mp/qrcode?scene=10000004&size=102&__biz=MjM5MzA0MzczMg==').then(result => {
  console.log(result)
})
```

node crashed with:

```
events.js:182
      throw er; // Unhandled 'error' event
      ^

Error: incorrect header check
    at Unzip.zlibOnError (zlib.js:152:15)
```

**Comment by puzrin** [comment]
Finally, short reproducible sample, thank you!

BTW, there is built-in script to try any url from console:

```sh
node support/dump.js https://mp.weixin.qq.com/mp/qrcode?scene=10000004&size=102&__biz=MjM5MzA0MzczMg==
[1] 7496
[2] 7497
vitaly@mbretina:~/Кодинг/probe-image-size$ events.js:182
      throw er; // Unhandled 'error' event
      ^

Error: incorrect header check
    at Unzip.zlibOnError (zlib.js:152:15)
```

Reopened issue as bug.

/cc @rlidwka

**Comment by puzrin** [comment]
Prior to apply more radical changes, i've reported [minimal sample](https://github.com/sindresorhus/got/issues/388#issuecomment-340141628) to got. May be they can fix now.

We could roll back to `request`, but it does not support retries.

**Comment by wong2** [comment]
@puzrin what about this retry lib? https://github.com/FGRibreau/node-request-retry

**Comment by wong2** [comment]
we've been using https://github.com/request/request-promise-native in production for a long time, and it works well. just fyi

**Comment by puzrin** [comment]
Promise interface is not ncessary. Prober needs stream + request interrupt.

**Comment by puzrin** [comment]
@wong2 is it acceptable if i roll back to `request` and "retry" feature will be lost?

**Comment by wong2** [comment]
@puzrin I think that's ok, the user can handle retry themselves if needed

**Comment by puzrin** [comment]
https://github.com/nodeca/probe-image-size/commit/c66e1147d5dbabf7c4539d4a716c851a926a02f4 Since we are busy and nothing moves forward, i've released 3.2.0 with reverted `got` dependency.

Plan is to release 4.0.0 with `request` (and without retries) in next weeks. `got` 8 was released with the same problems and it seems better to drop it in critical tasks with arbitrary url-s. Keep this ticket open until 4.0.0.

**Comment by wong2** [comment]
thanks

**Comment by Kikobeats** [comment]
one thing, consider use https://www.npmjs.com/package/simple-get for the future.

It's better than `request` because it's small and enough simple for your use case.

`got` was developed because `request` is very bloat.

**Comment by puzrin** [comment]
@Kikobeats what about api compatibility, request abort and retries? Could you explain details? Small size is nice fact, but not significant to decide.

-

**Comment by puzrin** [comment]
Released 4.0.0
